src/crystal/system/unix/process.cr

```diff
@@ -58,8 +58,34 @@ struct Crystal::System::Process
     raise RuntimeError.from_errno("kill") if ret < 0
   end
 
-  def self.on_interrupt(&handler : ->) : Nil
-    ::Signal::INT.trap { |_signal| handler.call }
+  def self.on_interrupt(&handler) : Nil
+    sig_handler = Proc(::Signal, Nil).new do |signal|
+      int_type = case signal
+                 when .int?
+                   ::Process::ExitReason::Interrupted
+                 when .hup?
+                   ::Process::ExitReason::TerminalDisconnected
+                 when .term?
+                   ::Process::ExitReason::SessionEnded
+                 else
+                   ::Process::ExitReason::Interrupted
+                 end
+
+      # maintain backwards compatibility
```

Comment:
```suggestion
      # NOTE: remove after crystal 1.12
      # maintain backwards compatibility
```

Reply:
would we be removing in 1.x? Are we adding in @straight-shoota 's suggestion for the compiler to accept Proc(X, Y) for a type restriction of Proc(Y)?

Reply:
That requires a separate discussion that has yet to take place so for now this change is not necessary.